query-service-impl/src/main/java/org/hypertrace/core/query/service/projection/ProjectionTransformation.java

```diff
@@ -341,4 +347,89 @@ private QueryRequest rebuildRequestOmittingDefaults(
+   * To handle order bys, we add the corresponding filter at the top and 'AND' it with the main filter.
+   * To handle filter, we modify each filter (say filter1) as : "CONTAINS_KEY AND filter1".
+   */
+  private Filter updateFilterForComplexAttributeExpression(
+      Filter filter, List<OrderByExpression> orderBys) {
+
+    Filter filterFromOrderBys = createFilterForComplexAttributeExpressionFromOrderBy(orderBys);
+    filter = updateFilterForComplexAttributeExpressionFromFilter(filter, filterFromOrderBys);
+
+    if (!filter.equals(Filter.getDefaultInstance())
+        && !filterFromOrderBys.equals(Filter.getDefaultInstance())) {
+      return Filter.newBuilder()
+          .setOperator(Operator.AND)
+          .addChildFilter(filter)
+          .addChildFilter(filterFromOrderBys)
+          .build();
+    } else if (!filterFromOrderBys.equals(Filter.getDefaultInstance())) {
+      return filterFromOrderBys;
+    } else {
+      return filter;
+    }
+  }
+
+  private Filter createFilterForComplexAttributeExpressionFromOrderBy(
+      List<OrderByExpression> orderByExpressionList) {
+
+    List<Filter> childFilterList =
+        orderByExpressionList.stream()
+            .filter(
+                orderByExpression ->
+                    orderByExpression.getExpression().getValueCase()
+                            == ValueCase.ATTRIBUTE_EXPRESSION
+                        && orderByExpression.getExpression().getAttributeExpression().hasSubpath())
+            .map(orderByExpression -> orderByExpression.getExpression().getAttributeExpression())
+            .map(
+                attributeExpression ->
+                    createContainsKeyFilter(
+                        attributeExpression.getAttributeId(), attributeExpression.getSubpath()))
+            .collect(Collectors.toList());
+
+    return childFilterList.isEmpty()
+        ? Filter.getDefaultInstance()
+        : Filter.newBuilder().setOperator(Operator.AND).addAllChildFilter(childFilterList).build();
+  }
+
+  private Filter updateFilterForComplexAttributeExpressionFromFilter(
+      Filter filter, Filter filterFromOrderBys) {
+
+    Filter.Builder builder = filter.toBuilder();
+    builder.clearChildFilter();
+
+    for (int i = 0; i < filter.getChildFilterCount(); i++) {
+      Filter childFilter =
+          updateFilterForComplexAttributeExpressionFromFilter(
+              filter.getChildFilterList().get(i), filterFromOrderBys);
+      builder.addChildFilter(childFilter);
+    }
+
+    filter = builder.build();
+
+    if (filter.getLhs().getValueCase() == ValueCase.ATTRIBUTE_EXPRESSION
+        && filter.getLhs().getAttributeExpression().hasSubpath()) {
+
+      Filter childFilter =
+          createContainsKeyFilter(
+              filter.getLhs().getAttributeExpression().getAttributeId(),
+              filter.getLhs().getAttributeExpression().getSubpath());
+
+      if (filterFromOrderBys.getChildFilterCount() > 0
```

Comment:
IMO, not even worth including this logic (duplicating the filter shouldn't cause any real issues, and for all we know this filter shows up somewhere else anyway), but if you want to include it, the `> 0` check is redundant, `contains` already covers that case.

Reply:
resolved